test/test_jit.py

```diff
@@ -1385,5 +1385,119 @@ def to_int(x) -> (y):
         outputs = Variable(torch.IntTensor([1, 2]), requires_grad=True)
         self.checkScript(script, 'to_int', [x], [outputs], False)
 
+    def test_script_while(self):
+        cu = torch.jit._jit_script_compile('''
+        def test_while(a, b) -> (c):
+            while a < 10:
+                a = a + 1
+                b = b + 1
+            c = a + b
+        ''')
+        self.assertExpected(str(cu.get_graph('test_while')))
+
+    def test_script_fibb(self):
+        cu = torch.jit._jit_script_compile('''
+        def test_while(lim) -> (third):
+            first = 1
+            second = 1
+            i = 1
+            somenum = 5
+            dontmutateme = 3
+            third = 0 # TODO: python lexical scoping
```

Comment:
If you have:

```
while False:
    third = ...
    ...
st = second + third
```

aren't you in trouble if you follow Python scoping? I am seriously skeptical initializing `third` to an undefined tensor here is a good idea.

Reply:
@zdevito had a good idea: let's do a simple data flow analysis to allow cases like:

```
if b:
   foo = True
else:
   foo = False
```

but not allow

```
if b:
   foo = ...
# now cannot prove that foo is always defined
```

Then we can have Python-style semantics at the frontend, but not have to bake in the semantics at the core level, because there you never need to model unbound variable access (as we've ruled it impossible by dataflow analysis.)